### ISSUE TITLE
dojox.widget.ColorPicker resource files urls

### ISSUE THREAD

**Issue by GordonSmith** [OpenNTF/dojo-webpack-plugin#64]
Packaging a web app which contains an instance of dojox.widget.ColorPicker.

The webpacked bundle ends up including the entire URL to the build folder and not the relative path, so once deployed the URLs cannot be resolved:
Actual:  ```http://192.168.3.22:8010/home/gordon/git/HPCC-Platform/esp/src/node_modules/dojox/widget/ColorPicker/images/pickerPointer.png```
Expected:  ```esp/src/node_modules/dojox/widget/ColorPicker/images/pickerPointer.png```

Looking at the code it appears to be a possible general issue with: ```require.toUrl("x/y/z.png")```

**Comment by chuckdumont** [comment]
`require.toUrl()` resolves URLs based on the value of the loaderConfig `baseUrl` property.  If you need `baseUrl` to have a different value on the client than it does during the build, you can do this by specifying the plugin's `loaderConfig` option as a string instead of as an object.  The string is the name of the loader config module (a CommonJS module) which will be loaded and evaluated separately during the build and during application load on the client.  You can then conditionally assign the `baseUrl` property depending on whether you are running in node or in a browser.

**Comment by chuckdumont** [comment]
To be more specific, `require.toUrl()`  uses the dojo config to determine the value of the URL.  This includes the `baseUrl`, `paths`, `packages`, etc.  Your specific error is for a resource in dojox.  Whether you can fix it by conditionally specifying `baseUrl` depends on how you're specifying the dojox location, if at all.  If you're specifying the location of dojox as a relative URL, then conditionally setting `baseUrl` would suffice.  Otherwise, you may also need to conditionally specify the `paths` and `packages` locations that you need to resolve correctly on the client. 

Note also that you'll need to provide the png as a stand-alone resource.  It cannot be packed in the application because it is loaded by the browser by URL.

**Comment by GordonSmith** [comment]
FYI - I tried to tweak the baseUrl, but couldn't coax it to build cleanly.  Since this was a one off issue in the entire project I went with the brute force approach of overriding the class and removing the require call:
```
    declare("HPCCColorPicker", [ColorPicker], {
        _underlay: "/esp/files/eclwatch/img/underlay.png",
        _hueUnderlay: "/esp/files/eclwatch/img/hue.png",
        _pickerPointer: "/esp/files/eclwatch/img/pickerPointer.png",
        _huePickerPointer: "/esp/files/eclwatch/img/hueHandle.png",
        _huePickerPointerAlly: "/esp/files/eclwatch/img/hueHandleA11y.png"
    });
```

**Comment by chuckdumont** [comment]
What kind of build errors did you run into?  Are you specifying the `loaderConfig` option as an object or as a module name?

**Comment by GordonSmith** [comment]
loaderConfig is a object for both dev and build "modes" and I didn't see how I could change it to a string without breaking the world - if you want to take a peek please do:
* https://github.com/GordonSmith/HPCC-Platform/blob/HPCC_JS/esp/src/webpack.config.js
* https://github.com/GordonSmith/HPCC-Platform/blob/HPCC_JS/esp/src/eclwatch/dojoConfig.js

The problem as I see it, is that the ```require.toUrl``` was burning the full URL (of the build folder) into the packaged files irrespective of what baseUrl was set to.

**Comment by chuckdumont** [comment]
Change the options you pass to DojoWebpackPlugin to

```javascript
		new DojoWebpackPlugin({
			loaderConfig: require.resolve("./js/loaderConfig"),
			environment: {dojoRoot: "node_modules"},
			locales: ["en"]
		}),
```

and then change the code in loaderConfig.js from

```javascript
function getConfig(dojoRoot) {
    var webRoot = "/esp/files";
```

to 

```javascript
function getConfig(env) {
    var dojoRoot = env.dojoRoot;
    var webRoot = "/esp/files";
```

Then add code to detect if you are running in a browser and adjust the value of dojoRoot to what you need for `require.toUrl()` to resolve correctly in the browser.

By loading the dojo loader config this way, the config values are not baked into the code at build time,  Instead, the loader config module is loaded and evaluated on the client, and you can do conditional processing based on the values of properties in the client.

I'm thinking of changing the sample to load the config this way since this approach is more flexible.

**Comment by GordonSmith** [comment]
Ahh that makes a lot of sense now (I was misreading which "loaderConfig" to change), it also means that a chunk of my refactoring is not needed (I moved a bunch of resolving code out of the loaderConfig into a diff init method in the bootstrap.js equivalent).  Thanks for taking time out to look at it.  I will make the changes and report back.

**Comment by chuckdumont** [comment]
I've added the [buildEnvironment](/OpenNTF/dojo-webpack-plugin#buildenvironment) option, which should make it easier to write a Dojo loader config that works as desired both at build time and at run-time.  This options is available in release 2.1.3 which is available on npm.
